minetest/minetest
# Unittest/collision

This PR adds unittests to help clarify issues with collision detection code (Issue 5966).

There are a set of classes in `unittest/mockgame.h` and `mockgame.cpp` to create a mock environment for conducting the tests. Each test sets up a map with a few nodes and tests one initial condition for collisionMoveSimple(). The final entity position is written to `rawstream` to aid with understanding why the test failed. 

## To do
* `unittest.cpp` also creates a mock game and makes it available to the unittests. The two mock game implementations should be merged.
* The test cases have extensive code in common. They should be refactored to improve maintainability.
* The current set of tests is optimized to highlight suspected problems found in a code review. The set of test cases should be expanded to identify more potential bugs. In particular, the tests only exercise motion in one direction.
* Some of the physics underlying the collision logic is not well-defined. I avoided a few test cases because I wasn't sure what the correct behavior ought to be.
* I have only tested on Windows compiled with `buildbot`.

This PR is Ready for Review.

## How to test
1. Ensure CMake is configured with `BUILD_UNITTESTS` enabled.
2. Build minetest. Install if necessary.
2. Open a console and change directory to where the minetest binary is located.
2. I recommend deleting or archiving the `../debug.txt` to make it easier to find the unittest results.
3. Execute `minetest --run-unittests`.
4. Inspect `../debug.txt`. Search for "Collision."  You should see 10 tests in this section, with 4 of them failing.
